BUG: check_api_dict does not correctly handle tuple values

eric-wieser opened the pull request:
It seems that at some point we augmented the indices with extra data. But this
data should not be included when determining uniqueness.

Previously, this would crash if there actually was a collision, rather than
printing a useful message

njsmith reviewed (commented):
Is it just me, or did the original code make no sense at all...? If we did have fully duplicate indices, wouldn't that be silently discarded before we ever made it into this function?

eric-wieser commented:
I ended up finding this because the code was detecting a duplicate index, but then crashed when trying to tell me which one it was. So the old code did work, and clearly didn't go with the transition to `(ind,)` over `ind`

charris commented on numpy/core/code_generators/genapi.py:
```diff
@@ -417,28 +417,33 @@ def merge_api_dicts(dicts):
-Same index has been used twice in api definition: %s
-""" % ['index %d -> %s' % (index, names) for index, names in doubled.items() \
-                                          if len(names) != 1]
+        msg = "Same index has been used twice in api definition: {}".format(
```
Might look better if this was broken into a `fmt` and `val`, and then `ValueError(fmt.format(val))` or some such.

eric-wieser replied:
Problem is, the format is sort of contained in `val` as well:

```
fmt = "Same index has been used twice in api definition: {}"
val = ''.join(
    '\n\tindex {} -> {}'.format(index, names)
    for index, names in doubled.items() if len(names) != 1
)
raise ValueError(fmt.format(val))
```


I don't think that `fmt` and `val` are useful names here, because both contain format strnigs

eric-wieser replied:
I guess I could do:
```
dupes = {index: names for index, names in doubled.items() if len(names) != 1}

raise ValueError("Same index has been used twice in api definition: {}".format(
    ''.join(
        '\n\tindex %d -> %s' % (index, names)
        for index, names in dupes.items()
    )
)
```

Although this is really bikeshedding, and I don't really want to switch working tree again...

charris replied:
Nothing wrong with bikeshedding, code clarity is part of the process...

eric-wieser replied:
Do you have a preference between the one in this PR and the two alternatives above?

charris replied:
I rather like the first. Probably put the closing `)` for `val` at the end of the preceding line there.

charris commented:
LGTM, maybe a style nit.

charris commented:
Need to finish this up.

eric-wieser committed:
BUG: check_api_dict does not correctly handle tuple values

It seems that at some point we augmented the indices with extra data. But this
data should not be included when determining uniqueness.

Previously, this would crash if there actually was a collision, rather than
printing a useful message

eric-wieser committed:
MAINT: Improve error message format

eric-wieser committed:
STY: Change string formatting line-wrapping

As requested by @charris

eric-wieser commented:
@charris: Done

eric-wieser commented:
@charris: Ping - ball is in your court now

charris commented:
Tnanks Eric.
